# How to do ssGSEA
*Raquel Manzano*


**Disclaimer: most text is taken from the [GSEA wiki](http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Main_Page), highly recommended if you want to know more and [original GSEA paper](https://www.pnas.org/content/102/43/15545). Here I put together what made me understand how GSEA works and how to use it.**

## Introduction
Gene Set Enrichment Analysis (GSEA) is a computational method that determines whether an *a priori* defined set of genes shows statistically significant, concordant differences between two biological states (e.g. phenotypes). Genes are ranked based on the correlation between their expression and the given phenotype. 

In other words, given *a priori* defined set of genes $S$, the goal of GSEA is to determine whether the members of $S$ are randomly distributed throughout ranked list $L$ or primarily found at the top or bottom. The score is calculated by walking down the list $L$, increasing a running-sum statistic when we encounter a gene in $S$ and decreasing it when we encounter genes not in $S$. The magnitude of the increment depends on the correlation of the gene with the phenotype.


The GSEA method was originally developed for analysis of microarray data. In order to best adapt this method for RNA-sequencing data sets the GSEA team has developed a [collection of guidelines and suggestions which describe how to properly handle these data](http://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Using_RNA-seq_Datasets_with_GSEA).


## Single-sample GSEA (ssGSEA)
If what you want to do is a *GSEA on a cohort without comparing phenotypes* you can do a single-sample GSEA (ssGSEA). Where GSEA generates a gene set’s enrichment score with respect to phenotypic differences across a collection of samples within a dataset, ssGSEA calculates a separate enrichment score for each pairing of sample and gene set, independent of phenotype labelling. In this manner, ssGSEA transforms a single sample's gene expression profile to a gene set enrichment profile. 

In other words, in ssGSEA each enrichment score represents the degree to which the genes in a particular gene set are coordinately up- or down-regulated within a sample. This transformation allows researchers to characterize cell state in terms of the activity levels of biological processes and pathways rather than through the expression levels of individual genes. In working with the transformed data, the goal is to find biological processes that are differentially active across the phenotype of interest and to use these measures of process activity to characterize the phenotype.


For the ssGSEA implementation, gene-level summed TPM serves as an appropriate metric for analysis of RNA-seq quantifications (to undertans RPKM, FPKM and TPM see this [fantastic post](https://rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/)).

To implemente ssGSEA into your analysis you will need two packages (follow links to install):

- [GSVA](https://bioconductor.org/packages/release/bioc/html/GSVA.html)
- [GSEABase](https://bioconductor.org/packages/release/bioc/html/GSEABase.html)

You can use the [Molecular Signature Database (MSigDB)](https://www.gsea-msigdb.org/gsea/msigdb/index.jsp) to download different gene sets related to specific pathways or diseases. E.g, if you click on C6 (oncogenic signatures) and then “Download GMT Files” you will obtain the file to do the analysis on the 189 gene sets that are related to cancer, you will need to provide your email for the download. There is a lot of information in the website, check it out.

If you want to try with the C6 collection here is the [download link](http://www.gsea-msigdb.org/gsea/msigdb/download_file.jsp?filePath=/msigdb/release/7.2/c6.all.v7.2.symbols.gmt)

### Simple example to run a ssGSEA
In this example I use data from [METABRIC](https://www.ncbi.nlm.nih.gov/pubmed/27161491,30867590,22522925), you can download the complete data set [here](http://download.cbioportal.org/brca_metabric.tar.gz). Once the download is complete you need to uncompress the file.

If you fancy using the terminal:
```{bash download_data eval=FALSE}
mkdir metabric_data  # created dir where data will be store
cd metabric_data  # go to this new dir
wget http://download.cbioportal.org/brca_metabric.tar.gz  # download data
tar -xvzf brca_metabric.tar.gz  # uncompress
```

This will download many files from metabric that you can use for different analysis but for now let's just use the z-scores. If you want to know more info on how cBioportal calculates these scores see [this link](https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats/z-score-normalization-script).


I extracted the Z-scores relative to all samples from the first 100 samples, which is in the input folder as data_mRNA_median_all_sample_Zscores.first100.txt.

```{r ssgsea}
library(GSVA)
library(GSEABase)
# read expression data
expr <- read.table('inputs/data_mRNA_median_all_sample_Zscores.first100.txt', 
                          sep='\t', header = T)

# format table as matrix, gsva function requires a matrix as input
rownames(expr) <- expr$Hugo_Symbol
expr$Hugo_Symbol <- NULL
expr$Entrez_Gene_Id <- NULL
expr.matrix <- as.matrix(expr)

# load gene set object
geneset <- getGmt("inputs/c6.all.v7.2.symbols.gmt")

# ssGSEA
res <- gsva(expr = expr.matrix,
            gset.idx.list = geneset,
            method = "ssgsea")

# simple heatmap of results
heatmap(res)

```